azkaban-common/src/main/java/azkaban/trigger/builtin/BasicTimeChecker.java

```diff
@@ -179,11 +226,30 @@ private long calculateNextCheckTime() {
           break;
         }
       } else {
+        // period is deprecated Schedule API parameter, we use cronSchedule API instead.
         date = date.plus(this.period);
       }
       count += 1;
     }
-    return date.getMillis();
+    // it's possible the above while loop exits on break condition instead of that "date refreshed after Now"
+    // this happens when trigger expires before NOW, in such case we remove all missed checkTimes as they're meaningless
+    if (!date.isAfterNow()) {
+      nextCheckTimeObj.missedCheckTimeBeforeNow.clear();
+    }
```

Comment:
Is it possible to have a few legit missed schedules before a trigger expiration? For example, let's say there is an schedule that should run 3 times: 1pm, 2pm and 3pm. The flow execution at 1pm ran as expected but then the Scheduler got stuck and the 2pm and 3pm runs were missed. The Webserver was restarted at 4pm, by then the schedule had already expired but there were 2 misses. Will the user be notified of those 2 schedule misses?

Reply:
That's an interesting question. This if only prevents expired triggers when recovered from DB, but does not prevent the expiring condition as you describe. 
But I think it currently needs a debate as Adam points that we should not have an expire cron.

Reply:
Got it, let's follow up on this then.